src/Sentry/Internal/CachingBackgroundWorker.cs

```diff
@@ -0,0 +1,146 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Sentry.Extensibility;
+using Sentry.Protocol.Envelopes;
+
+namespace Sentry.Internal
+{
+    internal class CachingBackgroundWorker : BackgroundWorkerBase
+    {
+        private const string EnvelopeFileExt = "envelope";
+        private readonly DirectoryInfo _cacheDirectory;
+
+        public CachingBackgroundWorker(ITransport transport, SentryOptions options)
+            : base(transport, options)
+        {
+            _cacheDirectory = !string.IsNullOrWhiteSpace(options.CacheDirectoryPath)
+                ? _cacheDirectory = new DirectoryInfo(options.CacheDirectoryPath)
+                : throw new InvalidOperationException("Cache directory is not set.");
+        }
+
+        private IEnumerable<FileInfo> GetEnvelopeFiles() =>
+            _cacheDirectory.Exists
```

Comment:
Should we check if it exists in the `ctor` and do some LogWarn or LogError?
I'm afraid this will be hard to debug

Reply:
That would probably be a bad idea. File system is very volatile, so even if it existed during construction, it may get deleted somehow after that, but before writing files. We should always treat working with FS as if it's in a constantly unpredictable state.

Reply:
I agree with everything you said. But I would expect to see some logs if the directory isn't there when the user explicitly set a path and expects offline caching to be working,

Reply:
Do you think it's unusual for someone to set a caching directory to something that doesn't exist? I personally would expect it to be normal.

I can add a log when we save a file, but the directory is not there.